src/components/KCodeBlock/KCodeBlock.vue

```diff
@@ -0,0 +1,968 @@
+    shouldPreventDefaultAction: true,
+  },
+
+  jumpToPreviousMatch: {
+    trigger: jumpToPreviousMatch,
+    isAllowedContext(event: Event) {
+      return codeBlock.value !== null && event.composedPath().includes(codeBlock.value)
+    },
+    isDisabled: () => matchingLineNumbers.value.length === 0 || isFilterMode.value,
+    shouldPreventDefaultAction: true,
+  },
+
+  copyCode: {
+    trigger: copyCode,
+    isAllowedContext(event: Event) {
+      return codeBlock.value !== null && event.composedPath().includes(codeBlock.value)
+    },
+  },
+}
+
+const KEY_REPLACEMENT: Record<string, string> = {
+  ' ': 'space',
+  Control: '',
+  Shift: '',
+  Alt: '',
+}
+
+function triggerShortcuts(event: KeyboardEvent): void {
+  const key = KEY_REPLACEMENT[event.key] !== undefined ? KEY_REPLACEMENT[event.key] : event.key.toLowerCase()
+  const shortcut = [
+    event.ctrlKey ? 'ctrl' : '',
+    event.shiftKey ? 'shift' : '',
+    event.altKey ? 'alt' : '',
+    key,
+  ].filter((key) => key !== '').join('+')
+  const commandKey = keyMap[shortcut]
+
+  if (!commandKey) {
+    return
+  }
+
+  const command = commands[commandKey]
+
+  // Prevents invoking shortcuts from outside a certain allowed context.
+  if (command.isAllowedContext) {
+    const isAllowedContext = command.isAllowedContext(event)
+
+    if (!isAllowedContext) {
+      return
+    }
+  }
+
+  if ((command.shouldPreventDefaultAction)) {
+    event.preventDefault()
+  }
+
+  if (command.isDisabled && command.isDisabled()) {
+    return
+  }
+
+  command.trigger(event)
+}
+
+function jumpToNextMatch(): void {
+  jumpToMatch(1)
+}
+
+function jumpToPreviousMatch(): void {
+  jumpToMatch(-1)
+}
+
+function jumpToMatch(direction: number): void {
+  if (matchingLineNumbers.value.length === 0 || !(codeBlock.value instanceof HTMLElement)) {
+    return
+  }
+
+  if (typeof currentLineIndex.value === 'number') {
+    currentLineIndex.value = ((currentLineIndex.value + direction) + matchingLineNumbers.value.length) % matchingLineNumbers.value.length
+  } else {
+    currentLineIndex.value = 0
+  }
+  const lineNumber = matchingLineNumbers.value[currentLineIndex.value]
+
+  if (!lineNumber) {
+    return
+  }
+
+  const line = codeBlock.value.querySelector(`#${linePrefix.value}-L${lineNumber}`)
+  if (line instanceof HTMLElement) {
+    if (typeof line.scrollIntoView === 'function') {
+      line.scrollIntoView({ block: 'center' })
+    }
+  }
+}
+
+function copyCode() {
+  copyTextToClipboard(props.code)
+}
+</script>
+
+<style lang="scss" scoped>
+.k-code-block {
+  --KCodeBlockBorderRadius: 5px;
+  --KCodeBlockFocusColor: var(--blue-500);
+  --KCodeBlockMatchHighlightColor: var(--blue-500);
+  --KCodeBlockColor: var(--black-85, color(black-85));
+  --KCodeBlockBackgroundColor: var(--grey-100);
+  --KCodeBlockFontSize: var(--type-xs);
+  --KCodeBlockFontFamilyMono: var(--font-family-mono);
+  --KCodeBlockTabSize: 2;
```

Comment:
If you define these variables, a user can never override them. Instead, you have to reference these CSS variables where they are used but then provide the values you have them set to as a fallback.

**You'll need to update all CSS variable usage like my example below**. Let me know if I can help

As an example:

```scss
.k-code-block pre,
.k-code-block code {
  tab-size: var(--KCodeBlockTabSize, 2);
  font-size: var(--KCodeBlockFontSize, var(--type-xs, 12px);
  font-family: var(--KCodeBlockFontFamilyMono, var(--font-family-mono, "Fira Mono", Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace));
  color: var(--KCodeBlockColor, var(--black-85, color(black-85)));
}
```

Reply:
TL;DR: I’ve changed this as suggested.

---

That deserves a bit more nuance.

It’s not true that set custom properties can’t be overridden. They can be overridden like any regular CSS property:

```css
.k-code-block.k-code-block {
  --KCodeBlockTabSize: 2;
}

.k-code-block {
  /* the computed value for `var(--KCodeBlockTabSize)` for elements matching the selector will be `2`. */
  --KCodeBlockTabSize: 4;
}
```

What _is_ true is that setting them here will prevent inheritance from rules that were defined earlier or which have selectors with lower specificity (somewhat likely because all Kongponents components unfortunately use scoped style blocks).

```css
:root {
  --KCodeBlockTabSize: 2;
}

.k-code-block {
   /* the computed value for `var(--KCodeBlockTabSize)` for elements matching the selector will be `4`. */
  --KCodeBlockTabSize: 4;
}
```

With the strategy of _not_ setting component-specific properties, you allow for globally defined default styles/themes which is good indeed. However, that comes at a cost: you can no longer specify a component’s default styles without getting very verbose like you suggested. Those defaults also can’t readily be controlled from one central place. Overall, I do think it’s a bigger benefit to allow for globally defined themes and not requiring to specifically set the theme on each component’s main selector (e.g. `.k-code-block { ... }`.

Reply:
> What _is_ true is that setting them here will prevent inheritance from rules that were defined earlier or which have selectors with lower specificity (somewhat likely because all Kongponents components unfortunately use scoped style blocks).

This is what I was referring to. A consuming app should be able to set these variables as is standard at the `:root {}` level instead of having to define them within the scope of the component classes.

> However, that comes at a cost: you can no longer specify a component’s default styles without getting very verbose like you suggested

If I understand your meaning, you still can. Instead of defining CSS variables within the component's `style` block, define them as Sass variables. Then you can map to both while still being concise:

```scss
$KCodeBlockBorderRadius: 5px;
$KCodeBlockFocusColor: var(--blue-500);
$KCodeBlockMatchHighlightColor: var(--blue-500);
$KCodeBlockColor: var(--black-85, color(black-85));
$KCodeBlockBackgroundColor: var(--grey-100);
$KCodeBlockFontSize: var(--type-xs);
$KCodeBlockFontFamilyMono: var(--font-family-mono);
$KCodeBlockTabSize: 2;

.k-code-block pre,
.k-code-block code {
  tab-size: var(--KCodeBlockTabSize, $KCodeBlockTabSize);
  font-size: var(--KCodeBlockFontSize, $KCodeBlockFontSize);
  font-family: var(--KCodeBlockFontFamilyMono, $KCodeBlockFontFamilyMono);
  color: var(--KCodeBlockColor, $KCodeBlockColor);
}
```

Reply:
That’s a good idea. I’ve defined the component’s own defaults like that now which reduces the verbosity a bit.

I’ve also noticed the missing imports:

```scss
@import '@/styles/variables';
@import '@/styles/functions';
```